protocol/statetracker/state_tracker.go

```diff
@@ -89,6 +101,31 @@ func (st *StateTracker) newLavaBlock(latestBlock int64, hash string) {
 	}
 }
 
+// block handler for emergency mode checking
+func (st *StateTracker) oldLavaBlock(latestBlock int64) {
+	st.registrationLock.RLock()
+	defer st.registrationLock.RUnlock()
+
+	latestBlockTime := st.EventTracker.getLatestBlockTime()
+	downtimeParams := st.chainTracker.GetDowntimeParams()
+
+	delay := time.Now().UTC().Sub(latestBlockTime)
+
+	// check if emergency mode is enabled
+	if delay < downtimeParams.DowntimeDuration {
+		return
+	}
+
+	epochDuration := downtimeParams.EpochDuration.Milliseconds()
+
+	// division delay by epoch duration rounded up, subtract one to skip regular epoch
+	virtualEpoch := (delay.Milliseconds()+epochDuration-1)/epochDuration - 1
+
+	for _, updater := range st.emergencyModeUpdaters {
+		updater.EmergencyModeUpdate(uint64(virtualEpoch))
```

Comment:
state tracker shouldnt be dealing with emergency updates, you need to create a new EmergencyTracker class to do it have it register for downtime params, have it be the source of truth for vitrtual epochs, have other classes embed an interface to it so you don't have to register for emergency updates